### PR TITLE
[WIP] Docker

### PR DESCRIPTION
**Pull request by Sindbag**
Dockerfile for production, devapi and cypress runs.

Also, environment settings for cypress and API urls

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -36,9 +37,10 @@
     "build": "webpack --config=webpack.prod.config.js",
     "test-serve": "concurrently -r \"yarn dev-api\" \"yarn server\"",
     "dev-api": "node ./dev-api",
-    "server": "cross-env DUCOR_FRONT_PORT=3090 ts-node server/index.ts",
+    "server": "ts-node server/index.ts",
```
Comment (zlumer) [inline package.json]: Remove `server` script completely with `server/index.ts`.
Node.js express server is now deprecated in favor of nginx.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+ARG DUCOR_FRONT_URL
+ARG DUCOR_FRONT_PORT
+ARG DUCOR_API_URL
+ARG DUCOR_API_PORT
+
+RUN npm run build --production
+
+# Stage 1, based on Nginx, to have only the compiled app, ready for production with Nginx
+FROM nginx:1.15
```
Comment (zlumer) [inline Dockerfile]: 👍

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+COPY --from=build-stage /app/build/ /usr/share/nginx/html
+# Copy the default nginx.conf provided by tiangolo/node-frontend
+COPY --from=build-stage /nginx.conf /etc/nginx/conf.d/default.conf
+EXPOSE 80
```
Comment (zlumer) [inline Dockerfile:18]: Port should be allowed to change from host machine by passing environment variable (or similar method)

File: Dockerfile-cypress [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# base image
```
Comment (zlumer) [inline Dockerfile-cypress]: Test-only environment is not necessary to run in Docker container, let's remove this image.

File: Dockerfile-devapi [hunk 5]
```diff
@@ -0,0 +1,34 @@
+FROM keymetrics/pm2:8-alpine
```
Comment (zlumer) [inline Dockerfile-devapi]: Test-only environment is not necessary to run in separate Docker container, let's remove this image.

File: cypress/integration/simple.spec.ts [hunk 6]
```diff
@@ -1,6 +1,6 @@
 describe('My First Test', () =>
 {
-	let URL = 'http://localhost:3090'
+	let URL = `http://${Cypress.env('URL') || 'localhost'}:${Cypress.env('PORT') || '80'}`
```
Comment (zlumer) [inline cypress/integration/simple.spec.ts]: Not necessary after removing Cypress-test image.

File: dev-api/index.js [hunk 7]
```diff
@@ -43,4 +43,4 @@ app
   .use(router.routes())
   .use(router.allowedMethods())
 
-module.exports = app.listen(3091)
+module.exports = app.listen(process.env.DUCOR_API_PORT)
```
Comment (zlumer) [inline dev-api/index.js]: Revert, never needed to change.

File: package.json [hunk 8]
```diff
@@ -36,9 +37,10 @@
...
     "test": "cypress run",
-    "ci": "yarn build && server-test test-serve \"3090|3091/config\" test"
+    "dotenv": "^6.0.0",
+    "ci": "yarn build && server-test test-serve \"$DUCOR_FRONT_PORT|$DUCOR_API_PORT/config\" test"
```
Comment (zlumer) [inline package.json]: Revert, not needed after removing Cypress-test image.

File: webpack.dev.config.js [hunk 9]
```diff
@@ -2,6 +2,19 @@ const path = require('path')
 const HtmlWebpackPlugin = require('html-webpack-plugin')
 const CleanWebpackPlugin = require('clean-webpack-plugin')
 const webpack = require('webpack')
+const dotenv = require('dotenv')
+
+const getEnvKeys = () => {
```
Comment (zlumer) [inline webpack.dev.config.js]: Why not [dotenv-webpack](https://github.com/mrsteele/dotenv-webpack)?

**Review by zlumer — CHANGES_REQUESTED** [review]
Let's stick to the following architecture:

1. Production docker image:
- builds app into static html with Webpack
- uses ENV vars to specify API `url` & `port` (and other variables)
- runs nginx on a specified port (not necessary 80)

2. Testing image:
- launches production docker container on e.g. port 3090; fails tests if webpack build fails
- launches dev-api (`yarn dev-api`)
- runs Cypress tests over production docker container

During development we will use `webpack-dev-server` for frontend development and run tests locally via `yarn dev-api` and `yarn test`.

File: Dockerfile (hunk 3 above)
Comment (Sindbag) [reply Dockerfile:18]: here we have default nginx.conf prod-ready from stage 0 contrainer, the port is hardcoded there, if we want to change it we have to maintain nginx.conf file
https://github.com/tiangolo/node-frontend/blob/master/nginx.conf

File: Dockerfile-cypress (hunk 4 above)
Comment (Sindbag) [reply Dockerfile-cypress]: So do we agree on cypress run in docker for travis ci or remove it

Comment (zlumer) [reply Dockerfile-cypress]: Let's keep it. Just patch the necessary functionality and add a testing docker-compose config.

File: webpack.dev.config.js [hunk 10]
```diff
@@ -3,6 +3,18 @@ const HtmlWebpackPlugin = require('html-webpack-plugin')
 const CleanWebpackPlugin = require('clean-webpack-plugin')
 const webpack = require('webpack')
 
+const getEnvKeys = () => {
```
Comment (zlumer) [inline webpack.dev.config.js]: As far as I can see, this code is no longer used.

**Review by zlumer — CHANGES_REQUESTED** [review]
Everything's okay, let's remove legacy code, fix `docker-compose down` and merge this.

**Review by zlumer — APPROVED** [review]
👍

**Pull request merged**